### ISSUE TITLE
As a traffic analyst I would like to have always real-time traffic and air quality data related to the A22

### ISSUE THREAD

**Issue by rcavaliere** [noi-techpark/bdp-commons#236]
CISMA is putting its models developed in BrennerLEC in production, but are complaining with the fact that the data is not always available in real-time. Currently the DC related to the innovative AQ sensors (MQTT) seems to be blocked. This user story should cover:
- [x] the quick solution of the issue with the innovative AQ sensors DC
- [x] a review of the logics so to ensure that we always have the freshest data available and a stable connection with the A22 systems

**Comment by bertolla** [comment]
My first impression of the problem is that the data collector is working correctly. The mqtt client embeded in the dc is the problem since after days of correct functioning it disconnects and does not reconnect. It might be a configuration issue or also a bug of the client itself. I will proceed by updating to the newest version of the tool and if it's still causing the same issue I will search for advanced configuration details. In the worst case we'll need to exchange the client.

**Comment by bertolla** [comment]
Redeployed the app just now, hoping that the auto-reconnect feature works: 
https://stackoverflow.com/questions/33735090/java-eclipse-paho-implementation-auto-reconnect

**Comment by rcavaliere** [comment]
I am not sure that is still working as expected. See the attached screenshot, if I compare Bluetooth data with traffic and air quality data related to A22... I think we have to a deeper look at the update logics in the correspondent Data Collector. Data should be updated with delays of minutes, not hours!
![Screenshot from 2021-03-02 21-39-44](https://user-images.githubusercontent.com/9825179/109712277-1f987100-7ba0-11eb-8784-7503d7597e1e.png)

**Comment by bertolla** [comment]
Oh! That's my bad. I made a miss configuration with the collector. It's starting right now. Should be working fine soon now

**Comment by rcavaliere** [comment]
Air quality sensors yes, traffic data can still be improved. I guess here we have some configurations in the elaborations done by Davide and Chris that have to be improved.
![Screenshot from 2021-03-03 21-40-11](https://user-images.githubusercontent.com/9825179/109869538-4b335e00-7c69-11eb-8780-6458eb765728.png)

**Comment by bertolla** [comment]
I've created a separated branch to test different mqtt configurations. Right now it's working correctly, but a lot of these settings were missing or completely wrong. I just ca do that wit trial and error right now, since logs don't give me lots of information.

**Comment by rcavaliere** [comment]
@bertolla currently the only sensor active is AQ3 since the other one are under calibration. This will last a couple of weeks and then the sensors will be installed again. I see the AQ3 sensor data as expected, so it seems now that the collector is stable in the real-time data flow. Let's leave this user story open in testing for a couple of additional days before to close it.

**Comment by rcavaliere** [comment]
@bertolla it can be that we have a new problem with the MQTT connector? I don't see data since this morning... see screenshot attached
![Screenshot from 2021-04-06 22-06-06](https://user-images.githubusercontent.com/9825179/113772026-6db30e00-9724-11eb-85af-dff412f0ad5a.png)

**Comment by bertolla** [comment]
please post the permalink not just the image

**Comment by bertolla** [comment]
@rcavaliere the data is here, it's just that your time range is not including today since default time for dates here is always 00:00:00

**Comment by bertolla** [comment]
For the traffic dc the problem lies with this query which takes a long time:
https://github.com/noi-techpark/bdp-commons/blob/3abe59e51111c8e022de7e5f85d11f475a0fa0e3/data-collectors/traffic-a22/a22elaborations/src/main/resources/it/bz/noi/a22elaborations/read-all-stations.sql

**Comment by bertolla** [comment]
After changing the query the data for traffic is up to date(takes around 10 minutes to do the calculations.) Following stations are not considered anymore for calculation after this change:
```
 A22:5416:1  | 2021-04-20 09:17:10
 A22:5416:2  | 2021-04-20 09:17:10
 A22:685:2   | 2021-04-20 09:17:10
 A22:685:1   | 2021-04-20 09:17:10
 A22:677:1   | 2021-04-20 09:17:10
 A22:685:3   | 2021-04-20 09:17:10
 A22:5416:4  | 2021-04-20 09:17:10
 A22:685:4   | 2021-04-20 09:17:10
 A22:5415:2  | 2021-04-20 09:17:10
 A22:890:2   | 2021-04-20 05:07:10
 A22:5438:1  | 2021-04-20 05:07:10
 A22:5438:2  | 2021-04-20 05:07:10
 A22:5438:3  | 2021-04-20 05:07:10
 A22:5438:4  | 2021-04-20 05:07:10
 A22:686:1   | 2021-04-20 05:07:10
 A22:686:2   | 2021-04-20 05:07:10
 A22:686:3   | 2021-04-20 05:07:10
 A22:686:4   | 2021-04-20 05:07:10
 A22:890:1   | 2021-04-20 05:07:10
```

**Comment by rcavaliere** [comment]
@bertolla can you please add the name of these stations? So I can check if they should be online or offline...

**Comment by bertolla** [comment]
I checked the values and they are all 0 and -999. Means they are offline. As long as we did calculation on all stations which were active in the last year there were done a lot of unneeded calculations.
The first one is `SEZIONE DI RILEVAMENTO KM. 076,600 RADAR (corsia di marcia nord, direzione nord)`. You can't see them on analytics, probably because `active` is set to `false`, which is the correct state.

**Comment by bertolla** [comment]
For the a22 environment elaborations(average) I noticed that the number of datapoints is often close to the min limit(30) to do the calculations, which means sometimes you won't see em because there are not 30 datapoints in that hour.

**Comment by rcavaliere** [comment]
@bertolla thanks for all these controls. OK for the elaborations as they are configured now, if the sensors do not measure something it's correct to not make any elaboration.

**Comment by bertolla** [comment]
For other datacollectors I saws there is no query which takes a long time. I noticed a regular 2h delay and think there must be some timezone issue on that one. Probably also on the others but I need to check

**Comment by bertolla** [comment]
for VMS it's the same

**Comment by bertolla** [comment]
and yes also for LinkStation.
The problem is in the datacollector for sure. 
@rcavaliere  Should I use even more time to investigate in the source code or should we pass it to Davide and Chris?

**Comment by rcavaliere** [comment]
I have seen that is much more efficient if you control this. If it takes you too long, please let me know.

**Comment by bertolla** [comment]
We might need to delete all old elaborations

**Comment by rcavaliere** [comment]
Here we don't have elaborations, what d you mean?

**Comment by bertolla** [comment]
I see, so these are raw data with wrong timestamps now. Do we get the old data?

**Comment by rcavaliere** [comment]
Yes of course. You intend to request the whole history since the timestamp was wrong? Can you fix this at a DB level?

**Comment by bertolla** [comment]
This is the issue:
![image](https://user-images.githubusercontent.com/3582457/115524333-4d3f9380-a28e-11eb-9372-3343b4fb79f2.png)

The code is not considering the timezone when parsing the date which in my eyes is wrong. There is an explanation here why https://github.com/noi-techpark/bdp-commons/blob/d802d0697241bf8c37c689fbd6d40090535f07d5/data-collectors/roadweather-a22/src/main/java/it/bz/noi/a22/roadweather/Connector.java#L376
but considering that we get the data with a 2 hour delay it's wrong or a22 is really sending old data.

I must also admit that the date format used by a22 is not following any convention I know, and it's really hard to interpret. Kind of borderline what they did there :disappointed_relieved: 

Since it's Davide's code I think he would be faster to fix this, as I found the issue.

**Comment by rcavaliere** [comment]
as you want. Do you want to notify this to Davide?

**Comment by bertolla** [comment]
> Yes of course. You intend to request the whole history since the timestamp was wrong? Can you fix this at a DB level?

No, we need to delete stuff that is in there or it would just be guessing.

**Comment by rcavaliere** [comment]
But we have already quite a long history of data... this could be quite demanding. Or?

**Comment by chris-mair** [comment]
Hi all,

the comment "Reverse engineering the A22 time stamp format" 

https://github.com/noi-techpark/bdp-commons/blob/d802d0697241bf8c37c689fbd6d40090535f07d5/data-collectors/roadweather-a22/src/main/java/it/bz/noi/a22/roadweather/Connector.java#L376

was made by me in the original A22 traffic connector.

I'm pretty sure I'm parsing the A22 timestamp correctly.

Indeed, if you look at my A22 traffic lag plots, you see the a22db lag is < 2 minutes (blue graph). 
I'll copy a recent plot here:

![a22](https://user-images.githubusercontent.com/4518241/115777974-3e540080-a3b6-11eb-8aaf-e65c0791bf7d.png)

Where the trouble starts is IMHO the BDP. There, the timestamp column ist stored WITTHOUT timezone, wich is a 
very troubled choice I've been recomending against for years :)

The whole problem is augmented by the fact for some reason the BDP is set to Europe/Rome as default zone, whereas all the servers are UTC. So I guess what happens is that if you guys try locally you see some time, when you query from the server you see another one.

TLDR: no, I don't think the original A22 traffic connector parses the timestamp wrongly. The long in the code fragment
@bertolla  hiliehted should be the correct epoch value.

The mess starts when Davide et.al. convert that epoch into a timestamp without timezone in the BDP. Again, sorry for
being blunt, but this was a design error in the BDP... 

I realize it is likely too late to fix now. So I suggest to bring the default timezone of everything in Ireland to the same zone (which would be UTC) as a first thing.

Then please everybody who coded a connector should try to do a max(ts) - now() to see if there are those two extra hours. Note that we're UTC + 2 when we have daylight saving and UTC + 1 in standard time. You should do this from a UTC environment to see what happens.

In the above graph you see the BDP (red) is lagging 2 hours which (after having fixed the other problem) ist clearly
a timezone issue. Probably it would go a away by simple by resetting the BDP default time zone to UTC to bring it in line
with the servers.

-- Chris.

**Comment by chris-mair** [comment]
Just to be perfectly clear, if I connect to the a22db right now, where those longs are stored as parsed by me,  I get:

a22db=> select max(timestamp) from a22.a22_traffic ;
    max     
------------
 1619123058
(1 row)

(there is no time zone conversion involved, were talking a long here, not a timestamp!)

And this value is indeed now!

UTC-Linux:

[chris@euler ~]$ date -d @1619123058
Thu Apr 22 20:24:18 UTC 2021

Europe/Rome-Linux:

chris@jupiter:~$ date -d @1619123058
Thu 22 Apr 2021 10:24:18 PM CEST

This is the correct time now.

So, again. The long is correct, you're messing it up when you store it or retrieve it from BDP.

-- Chris.

**Comment by chris-mair** [comment]
PS: in case this is not clear, the unix epoch is an *un*ambigous timestamp (it's the second since 1970-01-01 00:00:00 *UTC*).

-- Chris

**Comment by bertolla** [comment]
For the a22-traffic data collector I don't see any problems the newest dataset in the db is `A22:1871:1  | 2021-04-23 07:27:10 |    600`.
For the a22-roadweather data the newest is`A22:1871:1  | 2021-04-23 07:27:10 |    600`

**Comment by chris-mair** [comment]
@bertolla interesting, looks like different services have different behaviour :\

Here is a screenshot that confirms the long is already the correct epoch value for traffic:

![Screenshot 2021-04-23 at 10 05 21](https://user-images.githubusercontent.com/4518241/115850958-b6f2a580-a426-11eb-9f63-173285ff62ef.png)

I've also written a mail to you about this.

-- Chris.

**Comment by chris-mair** [comment]
Update:

it looks like after forcing the timestamp in the BDP to be interpretade as UTC there is no elaboration lag any more!

![a22](https://user-images.githubusercontent.com/4518241/116278260-9f991c80-a786-11eb-85eb-a6f0a572b15f.png)

So it least for traffic this should be solved ( I'd have @davidebz double check the elaboration code for time zone issues, though to be sure).

-- Chris

**Comment by rcavaliere** [comment]
Status: 
-> A22/traffic all solved, non dovrebbe esserci piu` nessun lag a nessun livello
-> A22/non-traffic pending ulteriori analisi Patrick Chris dopo input A22
-> A22 /air quality sensors: should work now

**Comment by bertolla** [comment]
- A22/RWISstations are now up to date(minutes) and should send valid data, timezone issues in history where also solved
-  A22/LinkStation are now up to date(minutes) and should send valid data, timezone issues in history where also solved
- A22/VMS are **not** up to date. The data provided by a22 was 2 hours behind before the timezone fix and are now 1 hour behind. Current example:
```
    timestamp      |       created_on        |                          string_value                          
---------------------+-------------------------+----------------------------------------------------------------
 2021-05-19 07:48:15 | 2021-05-19 08:51:55.069 | 1
 2021-05-19 07:48:15 | 2021-05-19 08:51:41.093 |  MOVIMENTO 
 2021-05-19 07:48:15 | 2021-05-19 08:51:41.128 | 32
 2021-05-19 07:48:15 | 2021-05-19 08:51:41.093 |  VORSICHT! 
 2021-05-19 07:47:45 | 2021-05-19 08:51:55.042 | 1
 2021-05-19 07:47:45 | 2021-05-19 08:51:40.951 | ATTENZIONE! LAVORI IN CORSO A 3 KM 
 2021-05-19 07:47:45 | 2021-05-19 08:51:41     | 32
 2021-05-19 07:47:45 | 2021-05-19 08:51:55.046 | 1
 2021-05-19 07:47:45 | 2021-05-19 08:51:55.042 | 2
 2021-05-19 07:47:15 | 2021-05-19 08:51:40.878 | ATTENZIONE! LAVORI IN CORSO A 7 KM 
 2021-05-19 07:47:15 | 2021-05-19 08:51:40.914 | 18
 2021-05-19 07:47:15 | 2021-05-19 08:51:55.037 | 1
 2021-05-19 07:47:15 | 2021-05-19 08:51:40.878 | VORSICHT! BAUSTELLE NACH 7 KM 
 2021-05-19 07:47:00 | 2021-05-19 08:51:55.017 | 2
 2021-05-19 07:47:00 | 2021-05-19 08:51:40.844 | 0
 2021-05-19 07:46:30 | 2021-05-19 08:51:54.994 | 1
 2021-05-19 07:46:30 | 2021-05-19 08:51:40.746 | ATTENZIONE! CODA DI MEZZI PESANTI 
 2021-05-19 07:46:30 | 2021-05-19 08:51:40.78  | 23
 2021-05-19 07:46:30 | 2021-05-19 08:51:40.746 | VORSICHT! LKW STAU
```

@rcavaliere Can you clarify with a22?

**Comment by rcavaliere** [comment]
@bertolla thanks for the update. The RWIS data seem to be complete. The LinkStation stations seem to have data only since 1.1.2020, is it right? Is the DC still ingesting data? See [here](https://analytics.opendatahub.bz.it/#%7B%22active_tab%22:0,%22height%22:%22400px%22,%22auto_refresh%22:false,%22scale%22:%7B%22from%22:1567288800000,%22to%22:1582844400000%7D,%22graphs%22:%5B%7B%22category%22:%22Road%20weather%22,%22station%22:%225507%22,%22station_name%22:%22KM%20119+600%20(L)%22,%22data_type%22:%22temp_suolo%22,%22unit%22:%22%5B%C2%B0C%5D%22,%22period%22:%221%22,%22yaxis%22:1,%22color%22:1%7D,%7B%22category%22:%22Vehicular%20time%22,%22station%22:%221854-1853%22,%22station_name%22:%22BOLZANO%20SUD%20-%20BOLZANO%20NORD%22,%22data_type%22:%22tempo%22,%22unit%22:%22%5Bs%5D%22,%22period%22:%221%22,%22yaxis%22:2,%22color%22:2%7D%5D%7D):
@chris-mair are we sure that the issue is not on our side, but on the A22 side?

**Comment by bertolla** [comment]
> @bertolla thanks for the update. The RWIS data seem to be complete. The LinkStation stations seem to have data only since 1.1.2020, is it right?

Seems to be right I also checked on the test-db where I didn't touch the data and this was the result of my query:
```
bdp=> select * from measurementhistory where station_id in (select id from station where stationtype ='LinkStation' and origin ='A22') 
and type_id in (select distinct type_id from measurement m join station s on s.id=m.station_id where stationtype ='LinkStation' and origin ='A22') and timestamp < '2020-01-01 00:00:00' order by timestamp desc;
 id | created_on | period | timestamp | double_value | provenance_id | station_id | type_id 
----+------------+--------+-----------+--------------+---------------+------------+---------
(0 rows)
```
> Is the DC still ingesting data?

Seem to be up to date:
select stationcode,name,timestamp from station join measurementstring on station_id=station.id where stationtype='LinkStation' and origin='A22';
 stationcode |               name                |      timestamp      
-------------+-----------------------------------+---------------------
 1840-1841   | CONFINE DI STATO - BRENNERO       | 2021-05-20 08:05:43
 1840-1841   | CONFINE DI STATO - BRENNERO       | 2021-05-20 08:05:43
 1840-1841   | CONFINE DI STATO - BRENNERO       | 2021-05-20 08:05:43
 1841-1842   | BRENNERO - TERME DI BRENNERO      | 2021-05-20 08:05:43
....
(156 rows)

**Comment by rcavaliere** [comment]
@bertolla @chris-mair I have seen that road weather data is not entering since 19.5. But the other data collectors seem to work (e.g. travel times). @chris-mair there is still another open point: @bertolla argues that VMS data are 1 hour late, and in fact it seems so. Before asking A22 (Federica), are we sure that on our side to read the data correctly?

**Comment by chris-mair** [comment]
Hi all,

please apologize, I opted out of mail from Gitlab so you need to send direct mail to reach me.

Here is my analysis.


First of all, checking with curl right now, it appears Federica's timestamp 
format has been reverted to the old one with the +0200 to be ignored! 

So again, the raw value we get is XXXXXXXXXX000+YYYY with XXXXXXXXXX being the
epoch and YYYY to be ignored.

All my collectors new or old accept XXXXXXXXXX000+YYYY or XXXXXXXXXX, so there
should be no problem.


-> traffico/transiti ("Traffic")

I'm long-term monitoring this.

Since May 4, no data is available between 0:00 and 2:00. I've checked the log:
the web service responds correctly with 0 records available. After 2:00, the
previously records from 0:00 to 2:00 become then available at once.

In other times of the day, the data is current.

See graph below.


-> meteo/misure ("Road Weather")

I'm not monitoring this, but right now trying with curl the data
is current. So, @rcavaliere if the data does not enter the BDP the
problem seems to be on our side.


-> /infoutenza/esposizioni ("PMV")

I have a hard time understanding if there is a time zone issue in the original
data just using curl, because the data is coming in sporadically and I'm not
long-term monitoring this.

I'm seeing what @bertolla showed. I agree there is a time zone issue,
but I cannot tell if it is upstream or it is caused by us running
the software that writes to the BDP in a TZ different from the one
BDP runs in. Since we usually run stuff either in UTC or in Europe/Rome 
(a difference of +2 now in summer) and @bertolla sees +1 and not +2, it 
might very well be the case this +1 comes from upstream. So I'm sort
of with @bertolla on this one.

-- Chris

![a22-traffic](https://user-images.githubusercontent.com/4518241/121033573-9b601600-c7ac-11eb-80d0-933a191dc74c.png)

**Comment by bertolla** [comment]
Regarding the roadweather data, there are some stations where data is up to date:
```
1888
2021
2022
```
while the others are still not receiving new data. I'll try to understand why.

**Comment by bertolla** [comment]
@rcavaliere 
Regarding VMS data there are 2 possibilities that come to my mind:
1. The data we receive are correct but are published one hour + some minutes later(strange since every other dataset is up to date)
2. A22 provides wrong timestamps (like they did for other datasets)

**Comment by bertolla** [comment]
I now also did some tests with curl and I get the same result.
There is also to note, that the pannels do not change periodically, but when they change we save it as a record.
My issue with the whole system is that no datarecord ever comes in prior one hour has passed, which means the most recent data is lagging always behind 1 hour and some minutes. Therefore the only possible reasons are the one I mentioned in the last comment and I tend strongly to number 2

**Comment by rcavaliere** [comment]
@bertolla can you write here the curl tests you tried for the VMS? I would like to investigate this more on myself. Please continue in the meantime to understand what's the issue with the RWIS data...

**Comment by bertolla** [comment]
Well you can but you will need to get a session token beforehand, configure the date parameters correctly and chenage the idpmv as you need, I'll send an encrypted session token that still works now:
```
curl -i -H 'Content-Type:application/json' -d '{"request":{"sessionId":"<token>","idpmv":5755,"fromData":"/Date(1623142600400+0000)/","toData":"/Date(1623143640999+0000)/"}}' "http://webservices.cau-a22.org:8080/A22Data/infoutenza/esposizioni
```

**Comment by bertolla** [comment]
I fixed the problem with the RWIS data.
For some strange reason, if a push to odh failed, the collector stopped and started from scratch.
Now it's not failing and even if one station fails to push it proceeds with the rest.
See 1c6f65c212420e8497e64e8229a6e3319618e7cf

**Comment by rcavaliere** [comment]
@chris-mair @bertolla For the VMS data: I had today another occasion to discuss this with Federica and CISMA, which is also reading automatically this data. They confirmed me that the data are provided now as **UTC + timezone**. CISMA needed to adapt their parser to correctly read the data. We should check if we are parsing correctly this. I forward you a Python application CISMA has implemented we can use for our checks.

**Comment by chris-mair** [comment]
@rcavaliere , we're moving in circles here...

F.C. already confirmed that the +0100 or +0200 is to be ignored. She even had it removed (read her email
below). For reasons unknown it appeared again.

All my code ignores the +0100 or +0200 (if it's present) and always has.

Appearantly right now we're off 1 hour for VMS only. If I added the +0200 that we're getting in summer,
we would be still wrong by 1 hour in the other direction for VMS and we would be off 2 hours for
everything else.

There's not so many ways to parse XXXXXXXXXX000+YY00. You can either take the XXXXXXXXXX
as epoch and ignore the YY or take the XXXXXXXXXX as epoch and add the YY hours.

Obviously the first way is the way (most of it) works.

The 1 hour off for VMS is either coming this way from upstream or is happening in the path from my code
to the BDP, but I'm inclined to believe @bertolla when he says he checked everything.

-- Chris

```
----- Forwarded message from "F******* C******" <********.*******@**********.**> -----
Date: 5/3/2021 at 5:13 PM
Subject: R: interpretazione date time
To: Chris Mair <chris@1006.org>
Cc: "*.*********@***.**.**" <*.*********@***.**.**>, ******* ********	<*.********@***.**.**>

Ciao Chris 
ho controllato la tua segnalazione, sicuramente c'è un errore sul server.
Il server ti invia sempre la data di misurazione in UTC (sia per i dati traffico che per i dati meteo) ma, errore nostro, non specifichiamo sul server che il kind della data è UTC.
Per questo la data viene presentata al client come se fosse in localtime, con +0100 o +0200 a seconda del fuso orario e dell'ora legale.
Modifico il servizio (installerò l'aggiornamento in serata o al massimo domani) e con questa modifica le date che riceverete saranno sempre formattate così  (es. “/Date(1521331219000)/”).

Per le date da specificare nelle ricerche invece, potete esprimerle in UTC oppure con il  fuso orario, predispongo il server per verificare se la data è in local o UTC e convertirla, se necessario.
Fammi sapere se con questa patch i dati vi risultano congruenti.

Ciao
F*******

-----Messaggio originale-----
Da: Chris Mair <chris@1006.org> 
Inviato: lunedì 26 aprile 2021 17:04
A: F******* C****** <********.*******@**********.**>
Cc: *.*********@***.**.**; ******* ******** <*.********@***.**.**>
Oggetto: interpretazione date time

Buongiorno,

di nuovo io :)

I web service del A22 ritornano datetime in questo formato qui:

“data”: “/Date(1521331219000+0100)/”

Da quando abbiamo iniziato a raccogliere i dati, io l'ho interpretato cosi`:

---*---

-> (token 1) 1521331219

e` lo unix epoch (definito come i secondi passati dal 1970-01-01 00:00:00 UTC), in questo caso 

Sun Mar 18 2018 00:00:19 UTC

equivalente a 

Sun Mar 18 2018 01:00:19 CET


-> (token 2) 000

sono i msec, che sono sempre 000 indicando che la precisione e` del secondo


-> (token 3) +0100

indica lo time zone in uso nel momento della misura, che, essendo di inverno era CET (+1) e non CEST (+2)

comunque il valore del (token 3) non influisce sull'interpretazione del timestamp, perche` il (token 1) da solo specifica un timestamp senza ambiguita`, essendo l'epoch per definizione riferito a UTC

---*---

A me questo sembra perfettamente coerrente perche` da anni sto ricevendo cosi` i dati di transito praticamente in tempo reale.

Ora pero` mi hanno solevato la questione che sembra che per altri servizi (tipo
meteo) sembra esserci un ritardo di tot ore che ci ha fatto venire il dubbio che ci sia un problema di timezone. 

Per questo ora stiamo tracciando tutti i passaggi e il primo, appunto, e` il parsing di questo date time.

Potresti per cortesia confermare la mia interpretazione esposta qui sopra (o casomai corregermi)? Inoltre possiamo assumere che sia consistente tra tutti i web service, giusto?

Grazie mille :)
Chris.
```

**Comment by chris-mair** [comment]
Hi,

In summary, this issue has become very broad, mixing different things.

Some issues effectly caused the data to be delayed:

- there were network (timeout) issues that stalled or blocked data gathering (confirmed and solved by us and A22)

- we had a performance issue with (some) elaborations that caused lags (confirmed and solved by us)

- some data was timestamped wrongly by A22 (confirmed and solved by A22)

Some other issued turned out to be non-issues, but added to the confusion:

- I was confused about the time zone in the BDP (this has been clarified to me by @bertolla) and made me wrongly assume the elaborations were delayed even more

- there was suspision about the timestamp format - I hope today's exchange of emails clarifies this once and for all: we DO NOT parse the timestamps wrongly

That being said, there appears to be one issue left, and that is that the PVM data seems to lag ~ 1 hour.

@rcavaliere: I suggest we clause this issue as resolved and open a new one about exactly that problem (~ 1 hour in PVM).

Actually there is another one, i.e. we get no data for traffic events between 00:00 and 02:00, but starting from 02:00 the missing data comes in. So you decide if this is an issue we should pursue with F.C. or if that's fine. I'm just mentioning it for completeness.

-- Chris.

**Comment by bertolla** [comment]
@rcavaliere 
With the last exchange of emails I understood that we interpret the data from vms the right way.
But I'm still conflicted. 
Is the vms service publishing the data 1 hour+ later after collecting the data?
We did not get any clear answer on that, or did I miss something?
And even if it's correct, what is the reason for the delay?

**Comment by rcavaliere** [comment]
@bertolla @chris-mair let us check this. Based on the answer we receive, let's close this issue definitely (so it looks like the data are correct) or let's close the issue and open a new one just for the VMS, as @chris-mair suggested
